packages/langium/src/workspace/documents.ts

```diff
@@ -304,18 +311,39 @@ export class DefaultLangiumDocuments implements LangiumDocuments {
         let langiumDoc = this.documentMap.get(uriString);
         if (langiumDoc) {
             // The document is already present in our map
-            return langiumDoc;
+            if (langiumDoc.state < DocumentState.Parsed) {
+                this.reloadDocumentContent(langiumDoc);
+            }
+
+        } else {
+            langiumDoc = this.loadDocumentContent(uri);
+            this.documentMap.set(uriString, langiumDoc);
         }
-        const textDoc = this.textDocuments.get(uriString);
+
+        return langiumDoc;
+    }
+
+    protected reloadDocumentContent(outdatedDocument: LangiumDocument) {
+        const newDocument = this.loadDocumentContent(outdatedDocument.uri);
+
+        // wipe out the existing document object...
+        Object.keys(outdatedDocument).forEach(key => delete outdatedDocument[key as keyof LangiumDocument]);
+        // ...and take over all the properties of the newly create one,
+        Object.assign(outdatedDocument, newDocument);
+        (outdatedDocument.parseResult.value as Mutable<AstNode>).$document = outdatedDocument;
```

Comment:
I'm not too happy with this solution, but since we need to keep the original document reference around, I guess there's no better solution? Another way would be to break the API of `LangiumDocumentFactory` to accept a document to override.

Reply:
Right. I also tried to add an optional function argument for replacing the call of `DefaultLangiumDocumentFactor.createLangiumDocument(...)` but that messed up the existing API considerably, so I skipped that.

Reply:
I agree with @msujew and propose to modify the properties explicitly:
 - Clear `references`, `precomputedScopes` and `diagnostics` in `invalidateDocument`
 - Also clear the TextDocument reference by using a "private" property instead of a local variable in `DefaultLangiumDocumentFactory.createLangiumDocument`; you can define such a property with a subinterface that is not exported
 - The `parseResult` is kept (see other discussion) and reassigned by the document builder